PR: https://github.com/odoo/odoo/pull/271273

From: d71d7b954ce3fa49b5c9d33129499b4deb0fc30a
From: Victor Feyens
Date: 2026-07-08 22:32:06

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = ["analytic.mixin", "res.currency.rate.consolidation.mixin"]

Total Changes: 255

[REF] product,account,*: simplify catalog logic

The catalog was historically developed for the `sale.order` (and `sale.order.line`)
models.

It was later on re-used for account moves and purchase orders, then
for stock models (mrp productions and work orders, ...).

In 19.0, it was enhanced to handle sections update (SO, PO, AM).

...

All those changes accumulated some inconsistencies and technical complexity.

This commit factorizes the logic duplicated in the different models as much as
possible, it also introduces a clean `product.catalog.line.mixin` that now
formalizes the methods expected on the 'line' model, and factorizes the default
behavior.

Note that the current commit (and PR) only focuses on the backend logic of the
catalog, the client side logic will require more work and will (hopefully) be done
in the near future.

Part-of: odoo/odoo#271273
Related: odoo/enterprise#121320
Related: odoo/upgrade#10572
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale/models/sale_order.py
+++ b/addons/sale/models/sale_order.py
@@ -2,7 +2,6 @@
 # ruff: noqa: PLW0642
 
 import json
-from collections import defaultdict
 from datetime import timedelta
 from itertools import groupby
 
@@ -1761,23 +1760,6 @@ class SaleOrder(models.Model):
         lines_to_recompute.discount = 0.0
         lines_to_recompute._compute_discount()
 
-    def _default_order_line_values(self, child_field=False):
-        default_data = super()._default_order_line_values(child_field)
-        new_default_data = self.env["sale.order.line"]._get_product_catalog_lines_data()
-        return {**default_data, **new_default_data}
-
-    def _get_action_add_from_catalog_extra_context(self):
-        return {
-            **super()._get_action_add_from_catalog_extra_context(),
-            "order_customer_id": self.partner_id.id,
-            "product_catalog_currency_id": self.currency_id.id,
-            "product_catalog_digits": self.order_line._fields["price_unit"].get_digits(self.env),
-            "show_sections": bool(self.id),
-        }
-
-    def _get_product_catalog_domain(self):
-        return super()._get_product_catalog_domain() & Domain("sale_ok", "=", True)
-
     @api.readonly
     def action_open_business_doc(self):
         self.ensure_one()
@@ -2135,9 +2117,7 @@ class SaleOrder(models.Model):
 
     def _discard_tracking(self):
         self.ensure_one()
-        return (
-            self.state == "draft" and request and request.env.context.get("catalog_skip_tracking")
-        )
+        return self.state == "draft" and request and request.env.context.get("catalog_update")
 
     def _track_finalize(self):
         """Override of `mail` to prevent logging changes when the SO is in a draft state."""
@@ -2638,6 +2618,25 @@ class SaleOrder(models.Model):
 
     # === CATALOG === #
 
+    def _get_action_add_from_catalog_extra_context(self):
+        return {
+            **super()._get_action_add_from_catalog_extra_context(),
+            "order_customer_id": self.partner_id.id,
+            "product_catalog_digits": self.order_line._fields["price_unit"].get_digits(self.env),
+        }
+
+    def _get_catalog_currency(self):
+        return self.currency_id or super()._get_catalog_currency()
+
+    def _get_product_catalog_domain(self):
+        return super()._get_product_catalog_domain() & Domain("sale_ok", "=", True)
+
+    def _get_product_price_type(self) -> str:
+        """Specify the price type that should be computed as product 'price' in the catalog."""
+        self.ensure_one()
+        # Disable the default price computation as the pricelist should be considered instead
+        return ""
+
     def _get_product_catalog_order_data(self, products, **kwargs):
         res = super()._get_product_catalog_order_data(products, **kwargs)
         prices = self.pricelist_id._get_products_price(
@@ -2651,91 +2650,33 @@ class SaleOrder(models.Model):
             res[product.id]["price"] = prices.get(product.id)
         return res
 
-    def _get_product_catalog_product_data(self, product, **kwargs):  # noqa: ARG002
-        product_data = super()._get_product_catalog_product_data(product)
+    def _is_readonly(self) -> bool:
+        """Return whether the sale order is read-only or not based on the state or the lock status.
+
+        A sale order is considered read-only if its state is 'cancel' or if the sale order is
+        locked.
+        """
+        return super()._is_readonly() or self.state == "cancel" or self.locked
+
+    def _get_product_catalog_product_data(self, product, **kwargs) -> dict:
+        product_data = super()._get_product_catalog_product_data(product, **kwargs)
         has_warning_group = self.env["res.groups"]._is_feature_enabled("sale.group_warning_sale")
         if product.sale_line_warn_msg and has_warning_group:
             product_data.update(warning=product.sale_line_warn_msg)
         return product_data
 
-    def _get_product_catalog_record_lines(self, product_ids, *, section_id=None, **kwargs):  # noqa: ARG002
-        grouped_lines = defaultdict(lambda: self.env["sale.order.line"])
-        if section_id is None:
-            section_id = (
-                self.order_line[:1].id
-                if self.order_line[:1].display_type == "line_section"
-                else False
-            )
-        for line in self.order_line:
-            if (
-                line.display_type
-                or line.product_id.id not in product_ids
-                or line.get_parent_section_line().id != section_id
-            ):
-                continue
-            grouped_lines[line.product_id] |= line
-        return grouped_lines
-
-    def _get_parent_field_on_child_model(self):
-        return "order_id"
-
-    def _update_order_line_info(
-        self, product, quantity, uom, *, section_id=False, child_field="order_line", **kwargs
-    ):
-        """Update sale order line information for a given product or create a
-        new one if none exists yet.
-
-        :param record product: The product, as a `product.product` record.
-        :param int quantity: The quantity selected in the catalog.
-        :param int section_id: The id of section selected in the catalog.
-        :param record uom: The UoM selected in the catalog, as a `uom.uom` record.
-        :return: The unit price of the product, based on the pricelist of the sale order and the
-                 quantity selected.
-        :rtype: float
-        """
-        request.update_context(catalog_skip_tracking=True)
-        sol = self.order_line.filtered(
-            lambda line: (
-                line.product_id.id == product.id and line.get_parent_section_line().id == section_id
-            )
+    def _get_product_catalog_default_unit_price(self, product, uom, **kwargs):
+        return self.pricelist_id._get_product_price(
+            product=product,
+            quantity=1.0,
+            currency=self.currency_id,
+            uom=uom,
+            date=self.date_order,
+            **kwargs,
         )
-        if sol:
-            if uom and sol.product_uom_id != uom:
-                sol.product_uom_id = uom.id
-            if quantity != 0:
-                sol.product_uom_qty = quantity
-            elif self.state in ["draft", "sent"]:
-                price_unit = self.pricelist_id._get_product_price(
-                    product=sol.product_id,
-                    quantity=1.0,
-                    currency=self.currency_id,
-                    uom=uom,
-                    date=self.date_order,
-                    **kwargs,
-                )
-                sol.unlink()
-                return price_unit
-            else:
-                sol.product_uom_qty = 0
-        elif quantity > 0:
-            sol = self.env["sale.order.line"].create({
-                "order_id": self.id,
-                "product_id": product.id,
-                "product_uom_qty": quantity,
-                "sequence": self._get_new_line_sequence(child_field, section_id),
-                "product_uom_id": uom.id,
-            })
-        else:  # quantity of 0, no line to update, return default pricelist price
-            return self.pricelist_id._get_product_price(
-                product=product,
-                quantity=1.0,
-                currency=self.currency_id,
-                uom=uom,
-                date=self.date_order,
-                **kwargs,
-            )
 
-        return sol._get_discounted_price()
+    def _has_sections(self) -> bool:
+        return True
 
     # === Product Documents === #
 
@@ -2758,18 +2699,6 @@ class SaleOrder(models.Model):
 
     # === TOOLING ===#
 
-    def _is_readonly(self):
-        """Return Whether the sale order is read-only or not based on the state or the lock status.
-
-        A sale order is considered read-only if its state is 'cancel' or if the sale order is
-        locked.
-
-        :return: Whether the sale order is read-only or not.
-        :rtype: bool
-        """
-        self.ensure_one()
-        return self.state == "cancel" or self.locked
-
     def _is_paid(self):
         """Return whether the sale order is paid or not based on the linked transactions.
 

--- a/addons/sale/models/sale_order_line.py
+++ b/addons/sale/models/sale_order_line.py
@@ -11,7 +11,11 @@ from odoo.tools import float_compare, float_is_zero, format_date, groupby
 
 class SaleOrderLine(models.Model):
     _name = "sale.order.line"
-    _inherit = ["analytic.mixin", "res.currency.rate.consolidation.mixin"]
+    _inherit = [
+        "analytic.mixin",
+        "res.currency.rate.consolidation.mixin",
+        "product.catalog.line.mixin",
+    ]
     _description = "Sales Order Line"
     _rec_names_search = ["name", "order_id.name"]
     _order = "order_id, sequence, id"
@@ -1798,13 +1802,35 @@ class SaleOrderLine(models.Model):
                 )
             )
 
-    # === ACTION METHODS ===#
+    # === CATALOG ===#
 
     @api.readonly
     def action_add_from_catalog(self):
         order = self.env["sale.order"].browse(self.env.context.get("order_id"))
         return order.with_context(child_field="order_line").action_add_from_catalog()
 
+    def _get_quantity_field(self) -> str:
+        return "product_uom_qty"
+
+    def _get_product_uom_field(self) -> str:
+        return "product_uom_id"
+
+    def _get_catalog_unit_price(self, parent_record, **kwargs):
+        if len(self) == 1:
+            return self._get_discounted_price()
+
+        return parent_record.pricelist_id._get_product_price(
+            product=self.product_id,
+            quantity=1.0,
+            uom=self._get_product_uom(),
+            currency=parent_record.currency_id,
+            date=parent_record.date_order,
+            **kwargs,
+        )
+
+    def _can_be_unlinked_from_catalog(self):
+        return super()._can_be_unlinked_from_catalog() and self.state in {"draft", "sent"}
+
     # === BUSINESS METHODS ===#
 
     def _expected_date(self):
@@ -2002,80 +2028,6 @@ class SaleOrderLine(models.Model):
     def _additional_name_per_id(self):
         return {so_line.id: so_line._get_partner_display() for so_line in self}
 
-    # === HOOKS ===#
-
-    def _get_product_catalog_lines_data(self, **kwargs):
-        """Return information about sale order lines in `self`.
-
-        If `self` is empty, this method returns only the default value(s) needed for the product
-        catalog. In this case, the quantity that equals 0.
-
-        Otherwise, it returns a quantity and a price based on the product of the SOL(s) and whether
-        the product is read-only or not.
-
-        A product is considered read-only if the order is considered read-only (see
-        ``SaleOrder._is_readonly`` for more details) or if `self` contains multiple records.
-
-        Note: This method cannot be called with records that have different products linked.
-
-        :raise odoo.exceptions.ValueError: ``len(self.product_id) != 1``
-        :rtype: dict
-        :return: A dict with the following structure:
-            {
-                'quantity': float,
-                'price': float,
-                'readOnly': bool,
-                'uomDisplayName': String,
-                'uomId': int,
-                'productUomFactor': float (optional),
-                'productUomDisplayName': string (optional),
-            }
-        """
-        if len(self) == 1:
-            available_uoms = self.product_id._get_available_uoms()
-            return {
-                "quantity": self.product_uom_qty,
-                "price": self._get_discounted_price(),
-                "readOnly": (self.order_id._is_readonly() or bool(self.combo_item_id)),
-                **self.order_id._get_product_catalog_uom_data(self.product_id, self.product_uom_id),
-                "availableUoms": [
-                    {"id": uom.id, "name": uom.display_name, "factor": uom.factor}
-                    for uom in available_uoms
-                ],
-            }
-        if self:
-            self.product_id.ensure_one()
-            order_line = self[0]
-            order = order_line.order_id
-            available_uoms = self.product_id._get_available_uoms()
-            return {
-                "readOnly": True,
-                "price": order.pricelist_id._get_product_price(
-                    product=order_line.product_id,
-                    quantity=1.0,
-                    currency=order.currency_id,
-                    date=order.date_order,
-                    **kwargs,
-                ),
-                "quantity": sum(
-                    self.mapped(
-                        lambda line: line.product_uom_id._compute_quantity(
-                            qty=line.product_uom_qty, to_unit=line.product_id.uom_id
-                        )
-                    )
-                ),
-                "uomDisplayName": self.product_id.uom_id.display_name,
-                "uomId": self.product_id.uom_id.id,
-                "availableUoms": [
-                    {"id": uom.id, "name": uom.display_name, "factor": uom.factor}
-                    for uom in available_uoms
-                ],
-            }
-        return {
-            "quantity": 0
-            # price will be computed in batch with pricelist utils so not given here
-        }
-
     # === TOOLING ===#
 
     def _convert_to_sol_currency(self, amount, currency):
